File: 2020/Day15/day15.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

#define LAST_TURN 30000000
std::vector<int> lst;
std::unordered_map<int, int> mem;


int main() {
    lst.reserve(LAST_TURN);
    int n;
    char _;
    while (std::cin >> n >> _)
        lst.push_back(n);
    lst.push_back(n);
    for (int i = 0; i < lst.size() - 1; i++)
        mem[lst[i]] = i;
    while (lst.size() != LAST_TURN) {
        int last = lst.back();
        auto itr = mem.find(last);
        if (itr == mem.end())
            lst.push_back(0);
        else
            lst.push_back(lst.size() - 1 - itr->second);
        mem[last] = lst.size() - 2;
    }
    std::cout << lst[2019] << '\n' << lst.back() << '\n';
}

File: 2020/Day15/day15.py
lst = list(map(int, input().split(',')))
mem = {x: i for i, x in enumerate(lst[:-1])}
while len(lst) != 30000000:
    last = lst[-1]
    i = mem.get(last)
    mem[last] = len(lst) - 1
    if i is None:
        lst.append(0)
    else:
        lst.append(len(lst) - 1 - i)
print(lst[2019], lst[-1], sep='\n')
